### ISSUE TITLE
cli: the `statement-diag` command doesn't clean up properly

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#67261]
**tldr** the `statement-diag` command does not clean up its bundle requests properly.

The `statement-diag` command is the only CLI command that uses the `sqlTxnShim`, an ancient mechanism that was intended to facilitate auto-retries in CLI code.

It turns out that `sqlTxnShim` is broken, and does not do what it says on the label.

To understand, look at the code:

```go
type sqlTxnShim struct {
  conn *sqlConn
}

var _ crdb.Tx = sqlTxnShim{}

func (t sqlTxnShim) Commit(_ context.context) error {
  return t.conn.Exec(`COMMIT`, nil)
}

func (t sqlTxnShim) Exec(...) error {
  // ... 
  return t.conn.Exec(query)
}

func (c *sqlConn) ExecTxn(fn func(*sqlConn) error) error {
  if err ≔ c.Exec(`BEGIN`); err != nil {
    return err
  }
  return crdb.ExecuteInTx(context.TODO(), sqlTxnShim{c}, func() error {
    return fn(c)
  })
}
```

This code assumes that the underlying SQL session where the `(t sqlTxnShim) Exec(...)` is executed is the same as the one `BEGIN` was executed under, and that all the calls to `(t sqlTxnShim) Exec(...)` are going to be executed atomically.

This is incorrect!  The method `(*sqlConn) Exec()` can detect TCP errors and automatically (under the hood) reestablishes a SQL connection if the connection drop.

So we can have the following sequence:

- caller calls ExecTxn() with callback, this issues a BEGIN
- callback calls Exec('DELETE'), this executed DELETE "under" the BEGIN
- connection drops
- callback calls Exec('INSERT'), this auto-reconnects and executes INSERT in a fresh connection

The problem is that `sqlTxnShim` delegates its `Exec` calls to `*sqlConn`. This should be fixed to delegate to the underlying `driver.Conn` instead.

cc @RaduBerinde @rytaft  for triage

**Comment by RaduBerinde** [comment]
There should be some common facility in the `cli` package for doing transactions. There are other usecases (in `renameUserFile`, `uploadFile`) and while those don't use the shim, they don't handle retries automatically either. Added SQL CLI project.

**Comment by knz** [comment]
the common facility *is* the sqlTxnShim and the `ExecInTxn()` method on the SQL conn object. Once we fix that, we can make the other things that need transactions use it.
